src/openms/include/OpenMS/FILTERING/TRANSFORMERS/SpectraMerger.h

```diff
@@ -297,22 +298,76 @@ namespace OpenMS
       exp.sortSpectra();
     }
 
+    /**
+     * @brief check if the first and second mzs might be from the same mass
+     *
+     * @param mz1 the first m/z value
+     * @param mz2 the second m/z value
+     * @param tol_ppm tolerance in ppm
+     * @param max_c maximum possible charge value
+     */
+    bool areMassesMatched(double mz1, double mz2, double tol_ppm, int max_c)
+    {
+      if (mz1 == mz2 || tol_ppm <= 0)
+      {
+        return true;
+      }
+
+      const int min_c = 1;
+      const int max_iso_diff = 5; // maximum charge difference  5 is more than enough
+      const double max_charge_diff_ratio = 3.0; // maximum ratio between charges (large / small charge)
+
+      for (int c1 = min_c; c1 <= max_c; ++c1)
+      {
+        double mass1 = (mz1 - Constants::PROTON_MASS_U) * c1;
+
+        for (int c2 = min_c; c2 <= max_c; ++c2)
```

Comment:
do you really need a quadratic algorithm to do this?
You can compute the charge (`c`) ratio from `mz1/mz2 == c2/c1`. The solution must be in this range. There should be no need to test them all...

Are the charges annotated really that unreliable? And even if not, maybe its worth testing them and see if it matches?

How much does the current solution slow down the algorithm?

Reply:
Currently this does not affect any running time on my macbook. But yes they can be improved as you mentioned. I will test with that as well. Very good advice.